src/workerd/api/pyodide/pyodide.c++

```diff
@@ -102,6 +116,176 @@ int ArtifactBundler::readMemorySnapshot(int offset, kj::Array<kj::byte> buf) {
+        }
+      }
+
+      return result;
+    };
+
+    // Skips one of the characters (specified in `oneOf`) at the current position. Otherwise
+    // throws. Returns the number of characters skipped.
+    auto skipChar = [](kj::StringPtr str, std::initializer_list<char> oneOf, int start) -> int {
+      for (char expected: oneOf) {
+        if (str[start] == expected) {
+          return 1;
+        }
+      }
+
+      KJ_FAIL_REQUIRE("Expected ", oneOf, "but received", str[start]);
+    };
+
+    auto parseKeyword = [](kj::StringPtr str, kj::StringPtr ident, int start) -> bool {
+      int i = 0;
+      for (; i < ident.size() && start + i < str.size(); i++) {
+        if (str[start + i] != ident[i]) {
+          return false;
+        }
+      }
+
+      return i == ident.size();
+    };
+
+    // Returns the size of the import identifier or 0 if no identifier exists at `start`.
+    auto parseIdent = [](kj::StringPtr str, int start) -> int {
+      // https://docs.python.org/3/reference/lexical_analysis.html#identifiers
+      //
+      // We also accept `.` because import idents can contain it.
+      if (std::isdigit(str[start])) {
+        return 0;
+      }
+      int i = 0;
+      for (; start + i < str.size(); i++) {
+        char c = str[start + i];
+        bool validIdentChar = std::isalpha(c) || std::isdigit(c) || c == '_' || c == '.';
+        if (!validIdentChar) {
+          return i;
+        }
+      }
+
+      return i;
+    };
+
+    int i = 0;
+    while (i < file.size()) {
+      auto keywordToParse = file[i] == 'i' ? "import"_kj : "from"_kj;
+      switch (file[i]) {
+        case ' ':
+        case '\t':
+          // Indented line, skip to newline.
+          i += skipUntil(file, {'\n', '\r'}, i);
+          if (file[i] != '\0') {
+            i += skipChar(file, {'\n', '\r'}, i);  // skip newline.
+          }
+          break;
+        case '\n':
+          // Windows users may submit \r\n newlines, block above would skip the `\r` so we'll be
+          // left with `\n` which we skip explicitly here.
+          i += 1;
+          break;
+        case 'i':
+        case 'f':
+          if (!parseKeyword(file, keywordToParse, i)) {
+            i++;
+            continue;
+          }
+          i += keywordToParse.size();  // skip "import" or "from"
+
+          while (i < file.size()) {
+            // Python expects a `\` to be paired with a newline, but we don't have to be as strict
+            // here.
+            i += skipWhile(
+                file, {'\r', '\n', ' ', '\t', '\\'}, i);  // skip whitespace and backslash.
+
+            if (file[i] == '.') {
+              // ignore relative imports
+              break;
+            }
+
+            int identLen = parseIdent(file, i);
+            KJ_REQUIRE(identLen > 0);
+
+            kj::String ident = kj::heapString(file.slice(i, i + identLen));
+            if (ident[identLen - 1] != '.') {  // trailing period means the import is invalid
```

Comment:
[nit] Might be nicer to handle where `.`s are and aren't allowed entirely within `parseIdent`?